src/value-indexer/src/main/scala/ValueIndexer.scala

```diff
@@ -0,0 +1,171 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import org.apache.hadoop.fs.Path
+import org.apache.spark.annotation.DeveloperApi
+import com.microsoft.ml.spark.schema._
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.util._
+import org.apache.spark.ml._
+import org.apache.spark.ml.attribute.NominalAttribute
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions.udf
+import org.apache.spark.sql.types._
+import reflect.runtime.universe.TypeTag
+
+import scala.math.Ordering
+import scala.reflect.ClassTag
+
+object ValueIndexer extends DefaultParamsReadable[ValueIndexer] {
+  def validateAndTransformSchema(schema: StructType, outputCol: String): StructType = {
+    val newField = NominalAttribute.defaultAttr.withName(outputCol).toStructField()
+    if (schema.fieldNames.contains(outputCol)) {
+      val index = schema.fieldIndex(outputCol)
+      val fields = schema.fields
+      fields(index) = newField
+      StructType(fields)
+    } else {
+      schema.add(newField)
+    }
+  }
+}
+
+/**
+  * Fits a dictionary of values from the input column.
+  * Model then transforms a column to a categorical column of the given array of values.
+  * Similar to StringIndexer except it can be used on any value types.
+  */
+class ValueIndexer(override val uid: String) extends Estimator[ValueIndexerModel]
+  with HasInputCol with HasOutputCol with MMLParams {
+
+  def this() = this(Identifiable.randomUID("ValueIndexer"))
+
+  /**
+    * Fits the dictionary of values from the input column.
+    *
+    * @param dataset The input dataset to train.
+    * @return The model for transforming columns to categorical.
+    */
+  override def fit(dataset: Dataset[_]): ValueIndexerModel = {
+    val dataType = dataset.schema(getInputCol).dataType
+    val levels = dataset.select(getInputCol).distinct().collect().map(row => row(0))
```

Comment:
Issues:
1. How to deal with missing values in the categorical columns?
2. How to deal with invalid values? Use the same behavior as StringIndexer or add an option to map invalid values to a "garbage" bin. 
3. Later: we need a way to handle merging of levels at some point. In AzureML, we would get datasets that were not properly schematized and ComputeModelStatistics needed to deal with mismatched levels
4. Much later: Need to specify an ordering function so distinct behaves deterministically on arbitrary values